### ISSUE TITLE
Sourcing `vim.lsp` results in redraw

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#27266]
### Problem

Plain sourcing of `vim.lsp` object results into redraw when not done directly during startup.

As one side effect, this results into intro text disappearing if user/plugin uses `vim.lsp` not during startup (for example, by setting LSP handler via `vim.lsp.handler`).

------

Bisecting 'vim.lsp' code to find the problem lead to following results:
- 'runtime/vim/lsp.lua' eagerly requires ['vim.lsp.semantic_tokens'](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp.lua#L9) and ['vim.lsp.inlay_hint'](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp.lua#L27).
- Both of them set decoration provider ([here](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp/semantic_tokens.lua#L795-L802) and [here](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp/inlay_hint.lua#L312-L358)) when sourced. Commenting out these two blocks fixes the issue.
- The `vim.api.nvim_set_decoration_provider()` [explicitly redraws "later"](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/src/nvim/api/extmark.c#L1048-L1049). @bfredl, mentioning you here as there is a `TODO(bfredl)` comment.

### Steps to reproduce

1. Create init file with the following content:
```lua
vim.schedule(function() local lsp = vim.lsp end)
```
2. `nvim --clean -u init.lua`. Intro text may be seen for split second but then is not shown.

This also happens with `nvim --clean` followed by `:lua require('vim.lsp')`.
This does not happen if 'init.lua' is plain `local lsp = vim.lsp`.

### Expected behavior

One of:
- Redraw to not happen at least when requiring `vim.lsp` or parts of it not related to decoration provider.
- Do not redraw or more smart redraw when setting decoration provider.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2243+g4ffc20c95

### Vim (not Nvim) behaves the same?

No

### Operating system/version

EndeavourOS Linux x86_64 (6.7.1-arch1-1)

### Terminal name/version

st-0.9

### $TERM environment variable

st-256color

### Installation

Both appimage and from source

**Comment by echasnovski** [comment]
I am not sure if this is a low complexity Lua issue. Postponing setting decoration provider for inlay hints and semantic tokens is not trivial (if even reasonably possible). So this relies on `vim.api.nvim_set_decoration_provider()` to not redraw immediately, which again might have unwanted side effects.

**Comment by gpanders** [comment]
>'runtime/vim/lsp.lua' eagerly requires ['vim.lsp.semantic_tokens'](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp.lua#L9) and ['vim.lsp.inlay_hint'](https://github.com/neovim/neovim/blob/4ffc20c9515294481486e81271a8edeeff203140/runtime/lua/vim/lsp.lua#L27).

Will this be addressed by [Dundarloading](https://github.com/neovim/neovim/pull/27216)?

**Comment by echasnovski** [comment]
> Will this be addressed by [Dundarloading](https://github.com/neovim/neovim/pull/27216)?

I've just tested that PR and I don't see an intro text disappearing. Neither with issue's 'init.lua', nor with slightly more complicated one:

```lua
vim.schedule(function()
  vim.lsp.handlers['$/progress'] = function() print('Hello') end
end)
```

So it fixes this particular issue but would imply that calling `vim.api.nvim_set_decoration_provider()` inside a script will cause a redraw (so should be used carefully).

**Comment by bfredl** [comment]
> So this relies on vim.api.nvim_set_decoration_provider() to not redraw immediately, which again might have unwanted side effects.

I think this should be allowed. it could be a flag to avoid changing the (for plugins actually reasonable) default.

**Comment by gpanders** [comment]
Fixed by https://github.com/neovim/neovim/pull/27852.

**Comment by justinmk** [comment]
> Fixed by #27852.

That PR made :intro more robust. But does requiring `vim.lsp` still cause a redraw?

**Comment by gpanders** [comment]
> > Fixed by #27852.
> 
> That PR made :intro more robust. But does requiring `vim.lsp` still cause a redraw?

No it does not, because of https://github.com/neovim/neovim/pull/27216. But this was only ever an issue in the first place because the redraw cleared the intro text (see https://github.com/echasnovski/mini.nvim/issues/640#issuecomment-1915587718).
